File: ios/AdBlock/PerformanceMetrics.swift
import Foundation

/// Performance metrics for the ad blocking engine
public struct PerformanceMetrics: Codable {
    let totalRequests: Int64
    let blockedRequests: Int64
    let allowedRequests: Int64
    let avgProcessingTimeNs: Int64
    let maxProcessingTimeNs: Int64
    let minProcessingTimeNs: Int64
    let filterCount: Int
    let memoryUsageBytes: Int
    let parseErrors: Int64
    let matchErrors: Int64
    let cacheHits: Int64
    let cacheMisses: Int64
    let cacheSize: Int
    let blockRate: Double
    let cacheHitRate: Double
    
    private enum CodingKeys: String, CodingKey {
        case totalRequests = "total_requests"
        case blockedRequests = "blocked_requests"
        case allowedRequests = "allowed_requests"
        case avgProcessingTimeNs = "avg_processing_time_ns"
        case maxProcessingTimeNs = "max_processing_time_ns"
        case minProcessingTimeNs = "min_processing_time_ns"
        case filterCount = "filter_count"
        case memoryUsageBytes = "memory_usage_bytes"
        case parseErrors = "parse_errors"
        case matchErrors = "match_errors"
        case cacheHits = "cache_hits"
        case cacheMisses = "cache_misses"
        case cacheSize = "cache_size"
        case blockRate = "block_rate"
        case cacheHitRate = "cache_hit_rate"
    }
    
    /// Get average processing time in microseconds
    var avgProcessingTimeMicros: Double {
        Double(avgProcessingTimeNs) / 1000.0
    }
    
    /// Get max processing time in microseconds
    var maxProcessingTimeMicros: Double {
        Double(maxProcessingTimeNs) / 1000.0
    }
    
    /// Get min processing time in microseconds
    var minProcessingTimeMicros: Double {
        Double(minProcessingTimeNs) / 1000.0
    }
    
    /// Get memory usage in MB
    var memoryUsageMB: Double {
        Double(memoryUsageBytes) / (1024.0 * 1024.0)
    }
    
    /// Format block rate as percentage string
    var formattedBlockRate: String {
        String(format: "%.1f%%", blockRate)
    }
    
    /// Format cache hit rate as percentage string
    var formattedCacheHitRate: String {
        String(format: "%.1f%%", cacheHitRate)
    }
    
    /// Create from JSON string
    static func fromJSON(_ json: String) -> PerformanceMetrics? {
        guard let data = json.data(using: .utf8) else { return nil }
        
        let decoder = JSONDecoder()
        return try? decoder.decode(PerformanceMetrics.self, from: data)
    }
}

File: android/app/src/main/java/com/adblock/model/PerformanceMetrics.kt
package com.adblock.model

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName

/**
 * Performance metrics data class
 */
data class PerformanceMetrics(
    @SerializedName("total_requests")
    val totalRequests: Long = 0,
    
    @SerializedName("blocked_requests")
    val blockedRequests: Long = 0,
    
    @SerializedName("allowed_requests")
    val allowedRequests: Long = 0,
    
    @SerializedName("avg_processing_time_ns")
    val avgProcessingTimeNs: Long = 0,
    
    @SerializedName("max_processing_time_ns")
    val maxProcessingTimeNs: Long = 0,
    
    @SerializedName("min_processing_time_ns")
    val minProcessingTimeNs: Long = 0,
    
    @SerializedName("filter_count")
    val filterCount: Int = 0,
    
    @SerializedName("memory_usage_bytes")
    val memoryUsageBytes: Int = 0,
    
    @SerializedName("parse_errors")
    val parseErrors: Long = 0,
    
    @SerializedName("match_errors")
    val matchErrors: Long = 0,
    
    @SerializedName("cache_hits")
    val cacheHits: Long = 0,
    
    @SerializedName("cache_misses")
    val cacheMisses: Long = 0,
    
    @SerializedName("cache_size")
    val cacheSize: Int = 0,
    
    @SerializedName("block_rate")
    val blockRate: Double = 0.0,
    
    @SerializedName("cache_hit_rate")
    val cacheHitRate: Double = 0.0
) {
    
    /**
     * Get average processing time in microseconds
     */
    fun getAvgProcessingTimeMicros(): Double {
        return avgProcessingTimeNs / 1000.0
    }
    
    /**
     * Get max processing time in microseconds
     */
    fun getMaxProcessingTimeMicros(): Double {
        return maxProcessingTimeNs / 1000.0
    }
    
    /**
     * Get min processing time in microseconds
     */
    fun getMinProcessingTimeMicros(): Double {
        return minProcessingTimeNs / 1000.0
    }
    
    /**
     * Get memory usage in MB
     */
    fun getMemoryUsageMB(): Double {
        return memoryUsageBytes / (1024.0 * 1024.0)
    }
    
    /**
     * Format block rate as percentage string
     */
    fun formatBlockRate(): String {
        return String.format("%.1f%%", blockRate)
    }
    
    /**
     * Format cache hit rate as percentage string
     */
    fun formatCacheHitRate(): String {
        return String.format("%.1f%%", cacheHitRate)
    }
    
    companion object {
        /**
         * Parse metrics from JSON string
         */
        fun fromJson(json: String): PerformanceMetrics? {
            return try {
                Gson().fromJson(json, PerformanceMetrics::class.java)
            } catch (e: Exception) {
                null
            }
        }
    }
}